oxd-server/src/main/java/org/gluu/oxd/server/service/KeyGeneratorService.java

```diff
@@ -0,0 +1,172 @@
+                JSONObject result = this.cryptoProvider.generateKey(algorithm,
+                        calendar.getTimeInMillis(), Use.ENCRYPTION);
+
+                JSONWebKey key = JSONWebKey.fromJSONObject(result);
+                jwks.getKeys().add(key);
+            } catch (Exception ex) {
+                LOG.error(ex.getMessage(), ex);
+            }
+        }
+
+        //LOG.trace("jwks: ", jwks);
+        LOG.trace("jwks generated successfully.");
+        return jwks;
+    }
+
+    public Jwt sign(Jwt jwt, String sharedSecret, SignatureAlgorithm signatureAlgorithm) {
+        try {
+            String signature = cryptoProvider.sign(jwt.getSigningInput(), jwt.getHeader().getKeyId(), sharedSecret, signatureAlgorithm);
+            jwt.setEncodedSignature(signature);
+            //return signed jwt
+            return jwt;
+        } catch (Exception e) {
+            LOG.error("Failed to sign signingInput.", e);
+            throw new RuntimeException("Failed to signingInput.", e);
+        }
+    }
+
+    public JSONWebKeySet getKeys() {
+        if (configuration.getEnableJwksGeneration()) {
+            if (keys != null && !keys.getKeys().isEmpty()) {
+                return this.keys;
+            }
+            //if keys not found then search in storage
+            JSONWebKeySet keys = getKeysFromStorage();
+            if (keys != null && !keys.getKeys().isEmpty()) {
+                this.keys = keys;
+                return this.keys;
+            }
+            //generate new keys in case they do not exist
+            generateKeys();
+            return this.keys;
+        }
+        LOG.info("Relying party JWKS generation is disabled in running oxd instance. To enable it set `enable_jwks_generation` field to true in `oxd-server.yml`.");
+        throw new HttpException(ErrorResponseCode.JWKS_GENERATION_DISABLE);
+    }
+
+    public void setKeys(JSONWebKeySet keys) {
+        this.keys = keys;
+    }
+
+    public String getKeyId(Algorithm algorithm, Use use) throws Exception {
+        try {
+            final String kid = cryptoProvider.getKeyId(getKeys(), algorithm, use);
+            if (!cryptoProvider.getKeys().contains(kid)) {
+                return cryptoProvider.getKeyId(getKeys(), algorithm, use);
+            }
+            return kid;
+
+        } catch (KeyStoreException e) {
+            LOG.error("Error in keyId generation");
+
+        }
+        return null;
+    }
+
+    public void saveKeysInStorage(String jwks) {
+        persistenceService.createExpiredObject(new ExpiredObject(ExpiredObjectType.JWKS.getValue(), jwks, ExpiredObjectType.JWKS, configuration.getJwksExpirationInHours() * 60));
+    }
+
+    public JSONWebKeySet getKeysFromStorage() {
+        ExpiredObject expiredObject = persistenceService.getExpiredObject(ExpiredObjectType.JWKS.getValue());
+        if (expiredObject != null && !Strings.isNullOrEmpty(expiredObject.getValue())) {
```

Comment:
Here before returning keys check whether its expired. And if expired remove it from storage.

Reply:
Done